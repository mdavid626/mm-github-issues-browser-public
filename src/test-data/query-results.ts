export const issueQueryResult = {
  data: {
    repository: {
      id: 'MDEwOlJlcG9zaXRvcnkxMDI3MDI1MA==',
      issue: {
        id: 'I_kwDOAJy2Ks5ePiI3',
        createdAt: '2023-02-12T05:31:50Z',
        title: 'Bug: build args parse error on powsershell',
        state: 'CLOSED',
        number: 26156,
        stateReason: 'COMPLETED',
        url: 'https://github.com/facebook/react/issues/26156',
        body: 'test issue result',
        author: { login: 'weidehai', __typename: 'User' },
        comments: {
          nodes: [],
          pageInfo: {
            hasNextPage: false,
            endCursor: null,
            __typename: 'PageInfo',
          },
          totalCount: 0,
          __typename: 'IssueCommentConnection',
        },
        __typename: 'Issue',
      },
      __typename: 'Repository',
    },
  },
};
