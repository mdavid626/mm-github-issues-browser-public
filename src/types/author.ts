export type Author = {
  login: string;
};
