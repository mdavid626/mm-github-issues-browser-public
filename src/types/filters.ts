export const stateFilters = ['open', 'closed'] as const;

export type StateFilter = (typeof stateFilters)[number];

export type Filters = {
  search: string;
  state: StateFilter | null;
};
