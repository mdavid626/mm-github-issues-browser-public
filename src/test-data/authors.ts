import { Author } from '../types/author';

export const testAuthor1: Author = {
  login: 'testAuthor1',
};
