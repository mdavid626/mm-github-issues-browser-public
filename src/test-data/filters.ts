import { Filters } from '../types/filters';

export const testFilters1: Filters = {
  search: 'test',
  state: 'open',
};
