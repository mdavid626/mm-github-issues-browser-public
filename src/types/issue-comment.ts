import { Author } from './author';
import { PageInfo } from './pagination';

export type IssueComment = {
  id: string;
  createdAt: string;
  body: string;
  author: Author;
};

export type IssueQueryComments = {
  nodes: IssueComment[];
  pageInfo: PageInfo;
  totalCount: number;
};
