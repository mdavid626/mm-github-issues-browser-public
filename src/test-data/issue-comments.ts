import { IssueComment, IssueComments } from '../types/issue-comment';
import { testAuthor1 } from './authors';

export const testIssueComment1: IssueComment = {
  id: 'IC_kwDOAJy2Ks5Ux_nZ',
  body: 'issue comment 1',
  createdAt: '2023-02-12T09:43:24.641Z',
  author: testAuthor1,
};

export const testIssueComment2: IssueComment = {
  id: 'IC_kwDOAJy2Ks5Ux4u7',
  body: 'issue comment 2',
  createdAt: '2022-01-11T10:42:25.655Z',
  author: testAuthor1,
};

export const testIssueComments1: IssueComments = {
  nodes: [testIssueComment1, testIssueComment2],
  pageInfo: {
    hasNextPage: true,
    endCursor: 'endCursor1',
  },
  totalCount: 2,
};

export const testEmptyIssueComments: IssueComments = {
  nodes: [],
  pageInfo: {
    hasNextPage: false,
    endCursor: 'endCursor1',
  },
  totalCount: 0,
};
