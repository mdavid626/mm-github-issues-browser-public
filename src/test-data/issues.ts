import { Issue, IssueQueryResult, IssuesItem } from '../types/issue';
import { testAuthor1 } from './authors';
import { testIssueComments1 } from './issue-comments';

export const testIssue1: Issue = {
  id: 'I_kwDOAJy2Ks5d65oG',
  createdAt: '2023-02-08T08:50:14Z',
  title: 'Bug: Failed when set homepage in package.json or set PUBLIC_URL',
  state: 'CLOSED',
  number: 26126,
  url: 'https://github.com/facebook/react/issues/26126',
  body: 'test issue1 body',
  author: testAuthor1,
  comments: testIssueComments1,
};

export const testIssuesItem1: IssuesItem = {
  id: 'I_kwDOAJy2Ks5d65oG',
  createdAt: '2023-02-08T08:50:14Z',
  title: 'Bug: Failed when set homepage in package.json or set PUBLIC_URL',
  state: 'CLOSED',
  number: 26126,
  url: 'https://github.com/facebook/react/issues/26126',
  author: testAuthor1,
};

export const testIssueQueryResult1: IssueQueryResult = {
  repository: {
    issue: testIssue1,
  },
};
