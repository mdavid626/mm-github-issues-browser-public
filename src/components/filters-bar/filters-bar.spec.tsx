import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { testFilters1 } from '../../test-data/filters';
import { Filters } from '../../types/filters';
import FiltersBar from './filters-bar';

describe('filters-bar', () => {
  afterEach(cleanup);

  it('should render', () => {
    const { asFragment } = render(
      <FiltersBar filters={testFilters1} setFilters={jest.fn()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render when no state filter', () => {
    const filters: Filters = {
      ...testFilters1,
      state: null,
    };
    render(<FiltersBar filters={filters} setFilters={jest.fn()} />);
    expect(screen.getByTestId('FiltersBar-state')).toHaveValue('');
  });

  it('should be able to select state filter', async () => {
    const filters: Filters = {
      ...testFilters1,
      state: 'open',
    };
    const setFilters = jest.fn();
    render(<FiltersBar filters={filters} setFilters={setFilters} />);
    await userEvent.selectOptions(
      screen.getByTestId('FiltersBar-state'),
      'Closed'
    );
    expect(setFilters).toHaveBeenCalledWith({
      ...filters,
      state: 'closed',
    });
  });

  it('should be able to select all state filter', async () => {
    const filters: Filters = {
      ...testFilters1,
      state: 'open',
    };
    const setFilters = jest.fn();
    render(<FiltersBar filters={filters} setFilters={setFilters} />);
    await userEvent.selectOptions(
      screen.getByTestId('FiltersBar-state'),
      'All'
    );
    expect(setFilters).toHaveBeenCalledWith({
      ...filters,
      state: null,
    });
  });

  it('should be able to search', async () => {
    const filters: Filters = {
      ...testFilters1,
      search: '',
    };
    const setFilters = jest.fn();
    render(<FiltersBar filters={filters} setFilters={setFilters} />);
    await userEvent.type(
      screen.getByPlaceholderText('Search...'),
      'test search'
    );
    await userEvent.click(screen.getByText('Search'));
    expect(setFilters).toHaveBeenCalledTimes(1);
    expect(setFilters).toHaveBeenCalledWith({
      ...filters,
      search: 'test search',
    });
  });

  it('should trigger search when pressing Enter in search box', async () => {
    const filters: Filters = {
      ...testFilters1,
      search: '',
    };
    const setFilters = jest.fn();
    render(<FiltersBar filters={filters} setFilters={setFilters} />);
    await userEvent.type(
      screen.getByPlaceholderText('Search...'),
      'test search{Enter}'
    );
    expect(setFilters).toHaveBeenCalledTimes(1);
    expect(setFilters).toHaveBeenCalledWith({
      ...filters,
      search: 'test search',
    });
  });

  it('should propagate search filter value to search text box when changes', () => {
    const filters: Filters = {
      ...testFilters1,
      search: 'test',
    };
    const { rerender } = render(
      <FiltersBar filters={filters} setFilters={jest.fn()} />
    );
    expect(screen.getByPlaceholderText('Search...')).toHaveValue('test');
    const updatedFilters: Filters = {
      ...testFilters1,
      search: 'new value',
    };
    rerender(<FiltersBar filters={updatedFilters} setFilters={jest.fn()} />);
    expect(screen.getByPlaceholderText('Search...')).toHaveValue('new value');
  });
});
