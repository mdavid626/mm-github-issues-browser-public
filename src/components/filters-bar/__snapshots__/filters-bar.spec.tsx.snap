// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters-bar should render 1`] = `
<DocumentFragment>
  <div
    class="FiltersBar"
  >
    <select
      class="FiltersBar-state"
      data-testid="FiltersBar-state"
    >
      <option
        value=""
      >
        All
      </option>
      <option
        value="open"
      >
        Open
      </option>
      <option
        value="closed"
      >
        Closed
      </option>
    </select>
    <input
      class="FiltersBar-search"
      placeholder="Search..."
      value="test"
    />
    <button
      class="FiltersBar-searchButton"
      type="button"
    >
      Search
    </button>
  </div>
</DocumentFragment>
`;
